monkeytype/typing.py

```diff
@@ -58,31 +88,47 @@ def __are_typed_dict_types_equal(type1: type, type2: type) -> bool:
 _TypedDictMeta.__eq__ = __are_typed_dict_types_equal
 
 
-def typed_dict_to_dict(type_dict: type) -> type:
-    """Convert a TypedDict to a Dict.
-    The keys of a TypedDict that we have constructed will always be strings.
-    However, if the TypedDict is empty, we return the key type as Any since we
-    can't justify a key type."""
-    if type_dict.__annotations__ == {}:
-        return Dict[Any, Any]
-    return Dict[str, shrink_types(type_dict.__annotations__.values())]
-
-
-def shrink_typed_dict_types(types: List[type]) -> type:
-    """Shrink the TypedDicts to a single TypedDict *only* if they are all the same.
-    Doing this manually because Union[TypedDict('Foo', {'a': int}),
-    TypedDict('Foo', {'a': int})] doesn't return a single TypedDict.
-
-    If there is any mismatch, fall back to a Union of the Dicts."""
-    assert len(types) != 0
-    first_type_dict = types[0]
-    if all(typ == first_type_dict for typ in types[1:]):
-        return first_type_dict
+def remove_requiredness_prefix(name: str) -> str:
+    if name.startswith(REQUIRED_FIELD_PREFIX):
+        prefix = REQUIRED_FIELD_PREFIX
+    elif name.startswith(NON_REQUIRED_FIELD_PREFIX):
+        prefix = NON_REQUIRED_FIELD_PREFIX
     else:
-        return Union[tuple(typed_dict_to_dict(typ) for typ in types)]
-
-
-def shrink_types(types):
+        prefix = ''
+    return name[len(prefix):]
+
+
+def shrink_typed_dict_types(typed_dicts: List[type], max_typed_dict_size: int) -> type:
+    """Shrink a list of TypedDicts into one with the required fields and the non-required fields.
+    Required fields are keys that appear in all the TypedDicts.
+    Non-required fields are those that appear in only some of the TypedDicts.
+    If the same key has a different value type, then its value is the Union of the value types."""
+    num_typed_dicts = len(typed_dicts)
+    key_value_types_dict = defaultdict(list)
+    for typed_dict in typed_dicts:
+        for key, value_type in typed_dict.__annotations__.items():
+            # Assumption: An input TypedDict to shrink_types will have all
+            # fields as required, since it was constructed directly from a
+            # dictionary object.
+            assert key.startswith(REQUIRED_FIELD_PREFIX)
```

Comment:
In this case, why not just skip the prefixes altogether in the TypedDicts constructed as input to this function?

(Though as mentioned above, I still think a custom class for internal representation of `TypedDict` may be a better approach.)